packages/protocol/contracts/L1/libs/LibAuction.sol

```diff
@@ -0,0 +1,275 @@
+
+    uint256 private constant ONE = 1_000_000;
+
+    event BatchBid(uint64 indexed batchId, uint64 startedAt, TaikoData.Bid bid);
+
+    error L1_BID_INVALID();
+    error L1_BATCH_NOT_AUCTIONABLE();
+    error L1_INSUFFICIENT_TOKEN();
+    error L1_INVALID_PARAM();
+    error L1_NOT_THE_BEST_BID();
+
+    function bidForBatch(
+        TaikoData.State storage state,
+        TaikoData.Config memory config,
+        uint64 batchId,
+        TaikoData.Bid memory bid
+    )
+        internal
+    {
+        if (
+            bid.prover != address(0) // auto-fill
+                || bid.blockMaxGasLimit != 0 // auto-fill
+                || bid.feePerGas == 0 || bid.proofWindow == 0 || bid.deposit == 0
+                || bid.proofWindow
+                    > state.avgProofWindow * config.auctionProofWindowMultiplier
+                || bid.deposit
+                    < state.feePerGas
+                        * (config.blockFeeBaseGas + config.blockMaxGasLimit)
+                        * config.auctionDepositMultipler
+        ) {
+            revert L1_BID_INVALID();
+        }
+
+        if (!isBatchAuctionable(state, config, batchId)) {
+            revert L1_BATCH_NOT_AUCTIONABLE();
+        }
+
+        bid.prover = msg.sender;
+        bid.blockMaxGasLimit = config.blockMaxGasLimit;
+
+        // Have in-memory and write it back at the end of the function
+        TaikoData.Auction memory auction =
+            state.auctions[batchId % config.auctionRingBufferSize];
+
+        // Deposit amount is per block, not per block * auctionBatchSize
+        uint64 totalDeposit = bid.deposit * config.auctionBatchSize;
+
+        if (batchId != auction.batchId) {
+            // It is a new auction
+            auction.startedAt = uint64(block.timestamp);
+            auction.bid = bid;
+            auction.batchId = batchId;
+            unchecked {
+                state.numAuctions += 1;
+            }
+        } else {
+            // An ongoing one
+            if (!isBidBetter(bid, auction.bid)) {
+                revert L1_NOT_THE_BEST_BID();
+            }
+            //'Refund' current
+            state.taikoTokenBalances[auction.bid.prover] += totalDeposit;
+        }
+
+        // Check if bidder at least have the balance
+        if (state.taikoTokenBalances[bid.prover] < totalDeposit) {
+            revert L1_INSUFFICIENT_TOKEN();
+        }
+
+        state.taikoTokenBalances[bid.prover] -= totalDeposit;
+        auction.bid = bid;
+
+        state.auctions[batchId % config.auctionRingBufferSize] = auction;
+
+        emit BatchBid(auction.batchId, auction.startedAt, bid);
+    }
+
+    function getAuctions(
+        TaikoData.State storage state,
+        TaikoData.Config memory config,
+        uint256 startBatchId,
+        uint256 count
+    )
+        internal
+        view
+        returns (uint256 currentTime, TaikoData.Auction[] memory auctions)
+    {
+        if (startBatchId == 0 || count == 0) {
```

Comment:
Do we need to also check `startBatchId` is not overflow / underflow the current ringbuffer?

Reply:
We don't need to. This function provides auctions for the client to verify (such as whether the auction is empty or not).